File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/CompositeMediaPlayer.java
package com.github.anrimian.musicplayer.data.controllers.music.players;

import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.error.ErrorType;
import com.github.anrimian.musicplayer.domain.models.player.events.ErrorEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;
import com.github.anrimian.musicplayer.domain.utils.functions.Function;

import javax.annotation.Nullable;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.subjects.BehaviorSubject;
import io.reactivex.rxjava3.subjects.PublishSubject;

public class CompositeMediaPlayer implements AppMediaPlayer {

    private final Function<AppMediaPlayer>[] mediaPlayers;
    private final int startPlayerIndex = 0;

    private final PublishSubject<PlayerEvent> playerEventSubject = PublishSubject.create();
    private final PublishSubject<Long> trackPositionSubject = PublishSubject.create();
    private final BehaviorSubject<Boolean> speedChangeAvailableSubject = BehaviorSubject.create();
    private final CompositeDisposable playerDisposable = new CompositeDisposable();

    private AppMediaPlayer currentPlayer;
    private int currentPlayerIndex;

    private long currentTrackPosition;

    private float currentPlaySpeed = 1f;

    @SafeVarargs
    public CompositeMediaPlayer(Function<AppMediaPlayer>... mediaPlayers) {
        this.mediaPlayers = mediaPlayers;

        setPlayer(startPlayerIndex);
    }

    @Override
    public Observable<PlayerEvent> getEventsObservable() {
        return playerEventSubject;
    }

    @Override
    public void prepareToPlay(CompositionSource composition,
                              long startPosition,
                              @Nullable ErrorType previousErrorType) {
        currentTrackPosition = startPosition;

        if (currentPlayerIndex != startPlayerIndex) {
            setPlayer(startPlayerIndex);
        }
        currentPlayer.prepareToPlay(composition, startPosition, previousErrorType);
    }

    @Override
    public void stop() {
        currentPlayer.stop();
    }

    @Override
    public void resume() {
        currentPlayer.resume();
    }

    @Override
    public void pause() {
        currentPlayer.pause();
    }

    @Override
    public void seekTo(long position) {
        currentPlayer.seekTo(position);
    }

    @Override
    public void setVolume(float volume) {
        currentPlayer.setVolume(volume);
    }

    @Override
    public Observable<Long> getTrackPositionObservable() {
        return trackPositionSubject;
    }

    @Override
    public Single<Long> getTrackPosition() {
        return currentPlayer.getTrackPosition();
    }

    @Override
    public Single<Long> seekBy(long millis) {
        return currentPlayer.seekBy(millis);
    }

    @Override
    public void setPlaySpeed(float speed) {
        this.currentPlaySpeed = speed;
        currentPlayer.setPlaySpeed(speed);
    }

    @Override
    public void release() {
        currentPlayer.release();
    }

    @Override
    public Observable<Boolean> getSpeedChangeAvailableObservable() {
        return speedChangeAvailableSubject;
    }

    private void setPlayer(int index) {
        currentPlayerIndex = index;
        if (currentPlayer != null) {
            currentPlayer.release();
        }
        currentPlayer = mediaPlayers[index].call();
        currentPlayer.setPlaySpeed(currentPlaySpeed);

        playerDisposable.clear();
        playerDisposable.add(currentPlayer.getEventsObservable()
                .flatMap(this::onPlayerEventReceived)
                .subscribe(playerEventSubject::onNext)
        );
        playerDisposable.add(currentPlayer.getTrackPositionObservable()
                .subscribe(this::onTrackPositionReceived)
        );
        playerDisposable.add(currentPlayer.getSpeedChangeAvailableObservable()
                .subscribe(speedChangeAvailableSubject::onNext));
    }

    private void onTrackPositionReceived(long position) {
        this.currentTrackPosition = position;
        trackPositionSubject.onNext(position);
    }

    private Observable<PlayerEvent> onPlayerEventReceived(PlayerEvent event) {
        return Observable.create(emitter -> {
            // if error event, switch to another player and consume event
            if (event instanceof ErrorEvent) {
                ErrorEvent errorEvent = ((ErrorEvent) event);
                ErrorType errorType = errorEvent.getErrorType();
                if (errorType == ErrorType.UNSUPPORTED) {
                    int newPlayerIndex = currentPlayerIndex + 1;
                    //don't switch player when we reached end of available players
                    if (newPlayerIndex >= 0 && newPlayerIndex < mediaPlayers.length) {
                        setPlayer(newPlayerIndex);
                        currentPlayer.prepareToPlay(errorEvent.getComposition(), currentTrackPosition, errorType);
                        return;
                    }
                }
            }
            emitter.onNext(event);
        });
    }
}

File: data/src/main/java/com/github/anrimian/musicplayer/data/controllers/music/players/CompositeMediaPlayer.kt
package com.github.anrimian.musicplayer.data.controllers.music.players

import com.github.anrimian.musicplayer.data.controllers.music.players.exceptions.PlayerOutOfMemoryException
import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.composition.content.RelaunchSourceException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnsupportedSourceException
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance
import com.github.anrimian.musicplayer.domain.models.player.events.MediaPlayerEvent
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.subjects.BehaviorSubject
import io.reactivex.rxjava3.subjects.PublishSubject

class CompositeMediaPlayer(
    private val mediaPlayers: ArrayList<() -> AppMediaPlayer>,
    private var currentPlaySpeed: Float,
    private var currentSoundBalance: SoundBalance,
    private var currentSkipSilenceEnabled: Boolean,
): AppMediaPlayer {

    private val playerEventsSubject = PublishSubject.create<MediaPlayerEvent>()
    private val playerDisposable = CompositeDisposable()

    private var currentPlayerIndex = START_PLAYER_INDEX
    private var nextPlayerIndex = currentPlayerIndex
    private var currentPlayer = createNewPlayerInstance(currentPlayerIndex)
    private val currentPlayerSubject = BehaviorSubject.createDefault(currentPlayer)

    private var currentSource: CompositionContentSource? = null
    private var previousPlayerException: Exception? = null

    override fun prepareToPlay(
        source: CompositionContentSource,
        previousException: Exception?,
    ): Completable {
        return Single.fromCallable {
            if (source == currentSource) {
                if (currentPlayerIndex != nextPlayerIndex) {
                    // switch player to next one
                    currentPlayerIndex = nextPlayerIndex
                    setPlayer(currentPlayerIndex)
                }
            } else {
                if (currentPlayerIndex == START_PLAYER_INDEX) {
                    if (nextPlayerIndex != START_PLAYER_INDEX) {
                        nextPlayerIndex = START_PLAYER_INDEX
                    }
                } else {
                    // reset player for new source
                    currentPlayerIndex = START_PLAYER_INDEX
                    setPlayer(currentPlayerIndex)
                }
            }
            currentPlayer
        }.flatMapCompletable { player ->
            //workaround for android media player unsupported error case
            var ex: Exception? = null
            if (source == currentSource) {
                ex = previousPlayerException
                previousPlayerException = null
            }
            currentSource = source
            player.prepareToPlay(source, ex)
        }.retry(this::prepareRelaunchOnError)
    }

    override fun stop() {
        currentSource = null
        currentPlayer.stop()
    }

    override fun resume() {
        currentPlayer.resume()
    }

    override fun pause() {
        currentPlayer.pause()
    }

    override fun seekTo(position: Long) {
        currentPlayer.seekTo(position)
    }

    override fun setVolume(volume: Float) {
        currentPlayer.setVolume(volume)
    }

    override fun getTrackPositionObservable(): Observable<Long> {
        return currentPlayerSubject.switchMap(AppMediaPlayer::getTrackPositionObservable)
    }

    override fun getTrackPosition(): Single<Long> {
        return currentPlayer.getTrackPosition()
    }

    override fun getDuration(): Single<Long> {
        return currentPlayer.getDuration()
    }

    override fun setPlaybackSpeed(speed: Float) {
        currentPlaySpeed = speed
        currentPlayer.setPlaybackSpeed(speed)
    }

    override fun setSkipSilenceEnabled(enabled: Boolean) {
        currentSkipSilenceEnabled = enabled
        currentPlayer.setSkipSilenceEnabled(enabled)
    }

    override fun release() {
        currentSource = null
        currentPlayer.release()
    }

    override fun getSpeedChangeAvailableObservable(): Observable<Boolean> {
        return currentPlayerSubject.switchMap(AppMediaPlayer::getSpeedChangeAvailableObservable)
    }

    override fun getPlayerEventsObservable(): Observable<MediaPlayerEvent> {
        return playerEventsSubject
    }

    override fun setSoundBalance(soundBalance: SoundBalance) {
        currentSoundBalance = soundBalance
        currentPlayer.setSoundBalance(soundBalance)
    }

    private fun setPlayer(index: Int) {
        currentPlayer.release()
        currentPlayer = createNewPlayerInstance(index)
        currentPlayerSubject.onNext(currentPlayer)
    }

    private fun createNewPlayerInstance(index: Int): AppMediaPlayer {
        val newPlayer = mediaPlayers[index]()
        newPlayer.setPlaybackSpeed(currentPlaySpeed)
        newPlayer.setSoundBalance(currentSoundBalance)
        newPlayer.setSkipSilenceEnabled(currentSkipSilenceEnabled)

        playerDisposable.clear()
        playerDisposable.add(newPlayer.getPlayerEventsObservable()
            .subscribe(this::onPlayerEventReceived)
        )
        return newPlayer
    }

    private fun onPlayerEventReceived(event: MediaPlayerEvent) {
        val eventToEmit = if (
            event is MediaPlayerEvent.Error && prepareRelaunchOnError(event.throwable)
        ) {
            MediaPlayerEvent.Error(RelaunchSourceException(event.throwable))
        } else {
            event
        }
        playerEventsSubject.onNext(eventToEmit)
    }

    private fun prepareRelaunchOnError(throwable: Throwable): Boolean {
        if (throwable is UnsupportedSourceException || throwable is PlayerOutOfMemoryException) {
            if (currentPlayerIndex >= 0 && currentPlayerIndex < mediaPlayers.lastIndex) {
                previousPlayerException = throwable as Exception
                nextPlayerIndex = currentPlayerIndex + 1
                return true
            }
        }
        return false
    }

    companion object {
        private const val START_PLAYER_INDEX = 0
    }

}